fix(compiler-cli): don't type check the bodies of control flow nodes in basic mode

crisbeto opened the pull request:
Angular only checks the contents of template nodes in full type checking mode. After v17, the new control flow always had its body checked, even in basic mode, which started revealing compilation errors for apps that were using the schematic to automatically switch to the new syntax.

These changes mimic the old behavior by not checking the bodies of `if`, `switch` and `for` blocks in basic mode. Note that the expressions of the blocks are still going to be checked.

Fixes #52969.

crisbeto commented on packages/compiler-cli/src/ngtsc/typecheck/api/api.ts:
```diff
@@ -335,6 +335,11 @@ export interface TypeCheckingConfig {
+  /**
+   * Whether to descend into the bodies of control flow blocks (`@if`, `@switch` and `@for`).
+   */
+  checkControlFlowBodies: boolean;
```
I was debating whether to reuse `checkTemplateBodies` instead of introducing this one. I decided to do it since it's a bit more explicit.

JoostK replied:
Yeah, I can go either way. I wouldn't personally have introduced it, but there's no harm in keeping it.

crisbeto commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1402,14 +1401,19 @@ class TcbIfOp extends TcbOp {
+    const checkBody = this.tcb.env.config.checkControlFlowBodies;
+    return Scope.forNodes(
+        this.tcb, parentScope, null, checkBody ? branch.children : [],
+        checkBody ? this.generateBranchGuard(index) : null);
```
Technically we don't need to skip the guard here since no children will be generated hence nothing will use it. I added the check here to avoid some of the work when we can.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts:
```diff
@@ -1747,6 +1794,22 @@ describe('type check blocks', () => {
+
+         const result = tcb(TEMPLATE, undefined, {checkControlFlowBodies: false});
+         expect(result).toContain('for (const _t1 of ((this).items)!) {');
+         expect(result).not.toContain('"" + ((this).main(_t1))');
+         expect(result).not.toContain('"" + ((this).empty())');
```
These are quite brittle as negative assertions, can we widen the matches to just `main` and `empty`, possibly with the preceding dot if that were to match undesirably.

JoostK reviewed (approved):
One request for changing a test, other than that LGTM.

crisbeto committed:
fix(compiler-cli): don't type check the bodies of control flow nodes in basic mode

Angular only checks the contents of template nodes in full type checking mode. After v17, the new control flow always had its body checked, even in basic mode, which started revealing compilation errors for apps that were using the schematic to automatically switch to the new syntax.

These changes mimic the old behavior by not checking the bodies of `if`, `switch` and `for` blocks in basic mode. Note that the expressions of the blocks are still going to be checked.

Fixes #52969.

alxhub commented:
This PR was merged into the repository by commit 7a16d7e969eaf5a9475ffdd21a4bf637ce523856.

minijus commented:
Could this fix be included in v17 as well?

crisbeto commented:
v17 port is in #55558.
